lib/puppet/util.rb

```diff
@@ -25,17 +25,47 @@ module Util
 
   # Run some code with a specific environment.  Resets the environment back to
   # what it was at the end of the code.
-  def self.withenv(hash)
-    saved = ENV.to_hash
-    hash.each do |name, val|
-      ENV[name.to_s] = val
-    end
+  # Windows can store unicode chars in the environment, so need to use API calls
+  #   to get and set the environment instead of ENV due to encoding problems.
+  def self.withenv(hash, mode=:posix)
```

Comment:
I don't agree.  If I replace the single if statement with a single begin-ensure, I end up with 3 if blocks doing different things.  If I move the functionality out to private helpers, I may as well take it further and withenv becomes a single if statement, branching out to private helpers called withenv_posix and withenv_windows and move ALL of the logic into there.  This can be done, as long as the yield statements have the same affect.